design-proposals/sdn-ipam-secondary-nets.md

```diff
@@ -0,0 +1,561 @@
+  loop for ipamClaim := range pod.ipamClaims
+  CNI->>CNI: configureIPs(pod, ipamClaim.ips)
+  end
+```
+
+**NOTES**:
+- When the feature gate is not enabled, KubeVirt will **not** request
+  persistent IP allocations from the CNI.
+- OVN-Kubernetes will **not** check for consistency errors related to the
+  `IPAMClaim`s used in two different pods simultaneously.
+
+#### Stopping a Virtual Machine
+```mermaid
+sequenceDiagram
+actor user
+
+user ->> KubeVirt: stopVM(vm=vmA)
+
+KubeVirt ->> Kubernetes: deletePod(owner=vmA)
+Kubernetes -->> KubeVirt: OK
+
+KubeVirt -->> user: OK
+
+note over OVN-K: notices a pod was deleted
+OVN-K ->> OVN-K: ips = teardownOVNEntities()
+```
+
+OVN-Kubernetes will **not** delete the `IPAMClaim` CRs for the deleted VM pod.
+The IP addresses allocated to the VM will **not** be released to the IP pool
+when there's a corresponding `IPAMClaim` for the attachment being deleted.
+
+#### Starting a (previously stopped) Virtual Machine
+This flow is - from a CNI perspective - quite similar to the
+[Creating a VM flow](#creating-a-virtual-machine):
+1. the workload controller (KubeVirt) templates the pod, featuring the required
+   `IPAMClaim` references in each network selection element.
+2. OVN-Kubernetes will read the network selection element; if it sees an
+   `IPAMClaim` reference, it will attempt to read it from the datastore.
+3. if the read `IPAMClaim` features IP addresses in its status, OVN-Kubernetes
+   will configure the interface with those IPs. If the status doesn't feature
+   IP addresses, the CNI plugin will retry later, with exponential backoff.
+
+As when creating a new VM, an error is thrown when the IPAMClaim belonged to an
+old VM with the same name. Eventually the garbage collector will remove this old
+IPAMClaim, and in the next reconcile loop the controller will create a new
+IPAMClaim.
+
+```mermaid
+sequenceDiagram
+
+  actor user
+  participant KubeVirt
+  participant Kubernetes
+  participant CNI
+
+  user->>KubeVirt: startVM(vmName)
+
+  note over KubeVirt: podName := "launcher-<vm name>"
+  KubeVirt->>Kubernetes: createPod(name=podName, ipam-claims=[claimName])
+
+  Kubernetes->>CNI: CNI ADD
+  note over CNI: this CNIs has an IPAMClaims informer
+
+  CNI->>CNI: ipamClaim := readIPAMClaim(claimName)
+  CNI->>CNI: configIface(ipamClaim.Status.IPs)
+
+  CNI-->>Kubernetes: OK
+
+  Kubernetes-->>KubeVirt: OK
+
+
+  KubeVirt-->>user: OK
+```
+
+When creating the pod, the OVN-Kubernetes IPAM module finds existing
+`IPAMClaim`s for the workload. It will thus use those already reserved
+allocations, instead of generating brand new allocations for the pod where the
+encapsulating object will run. The migration scenario is similar.
+
+#### Removing a Virtual Machine
+This flow is - from a CNI perspective - quite similar to the
+[Stopping a VM flow](#stopping-a-virtual-machine). The main difference is after
+the VM is deleted, Kubernetes Garbage Collection will kick in, and remove the
```

Comment:
In a foreground deletion -
1. The VM is marked to be removed.
2. Garbage collector tries to remove the IPAMClaim, since the IPAMClaim has a finalizer a deletion timestamp is added.
3. The VM is never removed since it has a dependent object that wasn't removed.
4. The IPAMClaim is not removed since the VM is not removed.

@oshoval can you please explain how does this work?

Reply:
Exactly as VM has a finalizer that is removed only when:
1. VM is marked for deletion
2. VMI is gone
when [1] + [2] happens, the VM finalizer is removed
this is the exact time when we also remove the Ipam claim finalizer
https://github.com/kubevirt/kubevirt/blob/e87c12294ae31810fdb75a6fb6f6b6d25f1f3107/pkg/virt-controller/watch/vm.go#L2974

Reply:
oh, I see. The finalizar is removed when the deletion timestamp is added. OK, thanks!